cli/azd/cmd/root.go

```diff
@@ -73,120 +76,147 @@ For more information, visit the Azure Developer CLI Dev Hub: https://aka.ms/azur
+	})
+
+	root.Add("show", &actions.ActionDescriptorOptions{
+		Command:        newShowCmd(),
+		FlagsResolver:  newShowFlags,
+		ActionResolver: newShowAction,
+		OutputFormats:  []output.Format{output.JsonFormat},
+		DefaultFormat:  output.NoneFormat,
+	})
+
+	root.Add("restore", &actions.ActionDescriptorOptions{
+		Command:        restoreCmdDesign(),
+		FlagsResolver:  newRestoreFlags,
+		ActionResolver: newRestoreAction,
+	})
+
+	root.Add("login", &actions.ActionDescriptorOptions{
+		Command:        newLoginCmd(),
+		FlagsResolver:  newLoginFlags,
+		ActionResolver: newLoginAction,
+		OutputFormats:  []output.Format{output.JsonFormat, output.NoneFormat},
+		DefaultFormat:  output.NoneFormat,
+	})
+
+	root.Add("logout", &actions.ActionDescriptorOptions{
+		Command:        newLogoutCmd(),
+		ActionResolver: newLogoutAction,
+	})
+
+	root.Add("monitor", &actions.ActionDescriptorOptions{
+		Command:        newMonitorCmd(),
+		FlagsResolver:  newMonitorFlags,
+		ActionResolver: newMonitorAction,
+	})
+
+	root.Add("down", &actions.ActionDescriptorOptions{
+		Command:        newDownCmd(),
+		FlagsResolver:  newDownFlags,
+		ActionResolver: newDownAction,
+		OutputFormats:  []output.Format{output.JsonFormat, output.NoneFormat},
+		DefaultFormat:  output.NoneFormat,
+	})
+
+	root.Add("init", &actions.ActionDescriptorOptions{
+		Command:        newInitCmd(),
+		FlagsResolver:  newInitFlags,
+		ActionResolver: newInitAction,
+	}).AddFlagCompletion("template", templateNameCompletion)
+
+	root.Add("up", &actions.ActionDescriptorOptions{
+		Command:        newUpCmd(),
+		FlagsResolver:  newUpFlags,
+		ActionResolver: newUpAction,
+		OutputFormats:  []output.Format{output.JsonFormat, output.NoneFormat},
+		DefaultFormat:  output.NoneFormat,
+	}).AddFlagCompletion("template", templateNameCompletion)
+
+	root.Add("provision", &actions.ActionDescriptorOptions{
+		Command:        newProvisionCmd(),
+		FlagsResolver:  newProvisionFlags,
+		ActionResolver: newInfraCreateAction,
+		OutputFormats:  []output.Format{output.JsonFormat, output.NoneFormat},
+		DefaultFormat:  output.NoneFormat,
+	})
+
+	root.Add("deploy", &actions.ActionDescriptorOptions{
+		Command:        newDeployCmd(),
+		FlagsResolver:  newDeployFlags,
+		ActionResolver: newDeployAction,
+		OutputFormats:  []output.Format{output.JsonFormat, output.NoneFormat},
+		DefaultFormat:  output.NoneFormat,
+	})
+
+	// Register any global middleware defined by the caller
+	if len(middlewareChain) > 0 {
+		for _, registration := range middlewareChain {
+			root.UseMiddlewareWhen(registration.Name, registration.Resolver, registration.Predicate)
 		}
 	}
 
-	return cmd
-}
+	// Global middleware registration
+	root.
+		UseMiddleware("debug", middleware.NewDebugMiddleware).
+		UseMiddlewareWhen("telemetry", middleware.NewTelemetryMiddleware, func(descriptor *actions.ActionDescriptor) bool {
+			return !descriptor.Options.DisableTelemetry
+		}).
+		UseMiddleware("ux", middleware.NewUxMiddleware)
```

Comment:
Slightly worried about `uxMiddleware`. The action result should be printed out that the end of middleware + action execution. This is not guaranteed by-design. For example, adding a command-level middleware would have it end up executing past the printout.

Reply:
FWIW - this also felt a little weird to me.

Reply:
Let me think about this one a little bit and get back to you all.

Reply:
Rolled back the move of this stuff into a middleware and added to the end of the root `RunE` command after root middleware chain + action is invoked.  We can revisit this later if needed.